### PR TITLE
Parallel mmult

### PR DESCRIPTION
**Pull request by Rombur**
In this PR, I've removed what was left of the `float` instantiation,  move code around to add a function that creates a `dealii::TrilinosWrappers::SparseMatrix` from a `SparseMatrixDevice`, and then use deal.II to do `mmult`. I have also moved around the destructor of `SparseMatrixDevice` because it was at the wrong place in the file but I haven't changed the code at all.

### CODE REVIEW

File: include/mfmg/sparse_matrix_device.cuh [hunk 1]
```diff
@@ -48,6 +48,9 @@ public:
 
   ~SparseMatrixDevice();
 
+  SparseMatrixDevice<ScalarType> &
+  operator=(SparseMatrixDevice<ScalarType> &&other);
```
Comment (dalg24) [inline include/mfmg/sparse_matrix_device.cuh]: Why are you providing a move assignment operator here?  You might want to do that in a separate commit.

File: include/mfmg/utils.cuh [hunk 2]
```diff
@@ -51,6 +51,9 @@ convert_matrix(dealii::TrilinosWrappers::SparseMatrix const &sparse_matrix);
 SparseMatrixDevice<double>
 convert_matrix(Epetra_CrsMatrix const &sparse_matrix);
 
+dealii::TrilinosWrappers::SparseMatrix
+convert_matrix(SparseMatrixDevice<double> const &matrix_dev);
```
Comment (dalg24) [inline include/mfmg/utils.cuh]: Same name for conversion `Epetra_CrsMatrix` -> `SparseMatrixDevice` and `SparseMatrixDevice` -> `dealii::TrilinosWrappers::SparseMatrix` is a confusing.

File: source/utils.cu [hunk 3]
```diff
@@ -167,6 +167,42 @@ SparseMatrixDevice<double> convert_matrix(Epetra_CrsMatrix const &sparse_matrix)
...
+  for (auto row : locally_owned_rows)
+  {
+    unsigned int const n_cols = row_ptr[pos + 1] - row_ptr[pos];
+    sparse_matrix.set(
+        row, n_cols,
+        reinterpret_cast<unsigned int *>(column_index.data() + row_ptr[pos]),
+        values.data() + row_ptr[pos]);
+    ++pos;
+  }
+  sparse_matrix.compress(dealii::VectorOperation::insert);
```
Comment (dalg24) [inline source/utils.cu]: Would you please add a comment saying that ideally we would populate directly the three vectors (row pointer, column indices, and values)?

Comment (Rombur) [reply source/utils.cu]: You can't do that though. There is no way to do that with deal, this is the best you can do.

File: include/mfmg/sparse_matrix_device.cuh (hunk 1 above)
Comment (Rombur) [reply include/mfmg/sparse_matrix_device.cuh]: Because in `mmult` I need to do `C = convert_matrix()` and I don't want to copy the matrix.

File: include/mfmg/sparse_matrix_device.templates.cuh [hunk 4]
```diff
@@ -341,38 +370,56 @@ void SparseMatrixDevice<ScalarType>::mmult(
...
+    ASSERT_CUSPARSE(cusparse_error_code);
+
+    // Reinitialize part of C
+    cuda_free(C.val_dev);
+    cuda_free(C.column_index_dev);
+    cuda_malloc(C.val_dev, C_local_nnz);
+    cuda_malloc(C.column_index_dev, C_local_nnz);
+    C._local_nnz = C_local_nnz;
+    C._nnz = C._local_nnz;
+    dealii::Utilities::MPI::sum(C._nnz, C._comm);
```
Comment (aprokop) [inline include/mfmg/sparse_matrix_device.templates.cuh]: You don't need it as `comm_size == 1`, right?

**Comment by aprokop** [comment]
(@Rombur next time, could you just push the commits addressing the comments on top of the branch without rebasing? That would make it easier to see how comments were addressed).

**Comment by Rombur** [comment]
Sure
